kernel/arch/dreamcast/kernel/entry.s

```diff
@@ -168,44 +154,31 @@ _save_regs_finish:
 	lds.l	@r1+,macl		! restore MACL 0x54	(+0x54)
 	ldc.l	@r1+,ssr		! restore SSR  0x58	(+0x58)
 
+	lds.l	@r1+,fpul		! restore FPUL  
 	mov	#0,r2			! Set known FP flags	(+0x5c)
 	lds	r2,fpscr
-	frchg				! Second FP bank
-	fmov.s	@r1+,fr0		! restore FR0  0x5c
-	fmov.s	@r1+,fr1		! restore FR1
-	fmov.s	@r1+,fr2		! restore FR2
-	fmov.s	@r1+,fr3		! restore FR3
-	fmov.s	@r1+,fr4		! restore FR4
-	fmov.s	@r1+,fr5		! restore FR5
-	fmov.s	@r1+,fr6		! restore FR6
-	fmov.s	@r1+,fr7		! restore FR7
-	fmov.s	@r1+,fr8		! restore FR8
-	fmov.s	@r1+,fr9		! restore FR9
-	fmov.s	@r1+,fr10		! restore FR10
-	fmov.s	@r1+,fr11		! restore FR11
-	fmov.s	@r1+,fr12		! restore FR12
-	fmov.s	@r1+,fr13		! restore FR13
-	fmov.s	@r1+,fr14		! restore FR14
-	fmov.s	@r1+,fr15		! restore FR15 0x98
-	frchg				! First FP bank
-	fmov.s	@r1+,fr0		! restore FR0  0x9c
-	fmov.s	@r1+,fr1		! restore FR1
-	fmov.s	@r1+,fr2		! restore FR2
-	fmov.s	@r1+,fr3		! restore FR3
-	fmov.s	@r1+,fr4		! restore FR4
-	fmov.s	@r1+,fr5		! restore FR5
-	fmov.s	@r1+,fr6		! restore FR6
-	fmov.s	@r1+,fr7		! restore FR7
-	fmov.s	@r1+,fr8		! restore FR8
-	fmov.s	@r1+,fr9		! restore FR9
-	fmov.s	@r1+,fr10		! restore FR10
-	fmov.s	@r1+,fr11		! restore FR11
-	fmov.s	@r1+,fr12		! restore FR12
-	fmov.s	@r1+,fr13		! restore FR13
-	fmov.s	@r1+,fr14		! restore FR14
-	fmov.s	@r1+,fr15		! restore FR15  0xd8
+	frchg                       ! Second FP bank
+    fschg                       ! Switch to pair FP moves
+    fmov.d  @r1+,dr0            ! restore FR0 & FR1
+    fmov.d  @r1+,dr2            ! restore FR2 & FR3
+    fmov.d  @r1+,dr4            ! restore FR4 & FR5
+    fmov.d  @r1+,dr6            ! restore FR6 & FR7
+    fmov.d  @r1+,dr8            ! restore FR8 & FR9
+    fmov.d  @r1+,dr10           ! restore FR10 & FR11
+    fmov.d  @r1+,dr12           ! restore FR12 & FR13
+    fmov.d  @r1+,dr14           ! restore FR14 & FR15
+    frchg                       ! First FP bank
+	add #32, r1
+!	fmov.d  @r1+,dr0            ! restore FR0 & FR1
+!   fmov.d  @r1+,dr2            ! restore FR2 & FR3
+!   fmov.d  @r1+,dr4            ! restore FR4 & FR5
+!   fmov.d  @r1+,dr6            ! restore FR6 & FR7
```

Comment:
I'm confused as to why you've got these commented out. It seems like a bad idea to not be restoring the entire state of all of the registers on return from an IRQ handler...